### PR TITLE
Implementation

### PR DESCRIPTION
**Pull request by melancholiai**

### CODE REVIEW

**Review by CptSchnitz — CHANGES_REQUESTED** [review]

File: src/commands/create/create.ts [hunk 1]
```diff
@@ -46,15 +79,34 @@ export class CreateCommand implements CommandModule<GlobalArguments, CreateArgum
...
+      if (error instanceof S3Error) {
+        exitCode = ExitCodes.S3_ERROR;
+      } else if (error instanceof HttpUpstreamUnavailableError) {
+        exitCode = ExitCodes.REMOTE_SERVICE_UNAVAILABLE;
+      } else if (error instanceof HttpUpstreamResponseError) {
+        exitCode = ExitCodes.REMOTE_SERVICE_RESPONSE_ERROR;
+      } else if (error instanceof DumpServerEmptyResponseError) {
+        exitCode = ExitCodes.DUMP_SERVER_EMPTY_RESPONSE_ERROR;
+      } else if (error instanceof Osm2pgsqlError) {
+        exitCode = ExitCodes.OSM2PGSQL_ERROR;
```
Comment (CptSchnitz) [inline src/commands/create/create.ts:101]: ?

File: src/commands/create/create.ts [hunk 2]
```diff
@@ -46,15 +79,34 @@ export class CreateCommand implements CommandModule<GlobalArguments, CreateArgum
...
-    const scriptPath = await this.manager.getScriptFromS3ToFs(s3BucketName, scriptKey);
+    try {
+      const scriptPath = await this.manager.getScriptFromS3ToFs(s3BucketName, scriptKey);
+
+      let dumpPath = dumpLocalFile;
+      if (source === DumpSource.REMOTE_URL) {
+        dumpPath = await this.manager.getDumpFromRemoteToFs(this.dumpSource as string);
+      } else if (source === DumpSource.DUMP_SERVER) {
+        dumpPath = await this.manager.getFromDumpServerToFs(this.dumpSource as string);
+      }
```
Comment (CptSchnitz) [inline src/commands/create/create.ts:87]: use same action for both options, just add a middle step

**Review by CptSchnitz — CHANGES_REQUESTED** [review]
separate helms to two

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      "internalConsoleOptions": "openOnSessionStart",
+      "cwd": "${workspaceFolder}/dist",
+      "skipFiles": ["${workspaceFolder}/node_modules/**/*.js", "<node_internals>/**/*.js"],
+      "outFiles": ["${workspaceFolder}/dist/**/*.js", "!**/node_modules/**"],
+      "program": "${workspaceFolder}/src/index.ts",
+      "env": {
+        "NODE_ENV": "development",
+        "SERVER_PORT": "8080",
+        "RESPONSE_COMPRESSION_ENABLED": "true",
+        "REQUEST_PAYLOAD_LIMIT": "1mb",
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply .vscode/launch.json]: done.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    && curl -L https://deb.nodesource.com/setup_${NODE_VERSION} | bash \
+    && apt-get -yq install nodejs libboost-filesystem-dev libpq-dev libproj-dev liblua5.3-dev libboost-program-options-dev
+
+COPY package*.json ./
+
+RUN npm ci --only=production
+
+COPY --from=buildApp /tmp/buildApp/dist .
+COPY --from=build /osm2pgsql/build /osm2pgsql
+COPY --from=build /osmium-tool/build /osmium-tool/build
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply Dockerfile]: done.

File: helm/templates/configmap.yaml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  POSTGRES_ENABLE_SSL_AUTH: {{ .sslAuth.enabled | quote }}
+  {{- if .sslAuth.enabled }}
+  POSTGRES_CERTS_MOUNT_PATH: {{ .sslAuth.mountPath | quote }}
+  {{- end }}
+  {{- end }}
+  {{- with .Values.s3 }}
+  S3_ENDPOINT: {{ printf "%s://%s:%v/" .protocol .host .port | quote }}
+  S3_BUCKET_NAME: {{ .bucketName | quote }}
+  S3_KEY_ID: {{ .keyId | quote }}
+  S3_ACL: {{ .acl | quote | default "public-read" }}
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply helm/templates/configmap.yaml]: removed default value.

File: helm/templates/command-options-configmap.yaml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  {{- if eq .Values.cli.command "append" }}
+  CONFIG: {{ printf "%s/%s" .Values.cli.append.config.mountPath .Values.cli.append.config.fileName | quote }}
+  REPLICATION_URL: {{ .Values.cli.append.replicationUrl | quote }}
+  {{- end }}
+  {{- if eq .Values.cli.command "create" }}
+  SOURCE: {{ .Values.cli.create.source | quote }}
+  S3_SCRIPT_KEY: {{ .Values.cli.create.s3ScriptKey | quote }}
+  DUMP_LOCAL_FILE: {{ .Values.cli.create.dumpLocalFile | quote }}
+  DUMP_REMOTE_URL: {{ .Values.cli.create.dumpRemoteUrl | quote }}
+  DUMP_SERVER_URL: {{ .Values.cli.create.dumpServerUrl | quote }}
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply helm/templates/command-options-configmap.yaml]: changed the create command options to have a `dump-source-type` and a `dump-source`, so this is no longer an issue

File: helm/templates/configmap.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  TELEMETRY_TRACING_URL: {{ $tracingUrl }}
+  {{- else }}
+  TELEMETRY_TRACING_ENABLED: "false"
+  {{- end }}
+  HTTP_CLIENT_TIMEOUT: {{ .Values.env.httpClient.timeout | quote }}
+  {{ if .Values.env.osm2pgsql.slim }}
+  OSM2PGSQL_SLIM: "true"
+  {{- else }}
+  OSM2PGSQL_SLIM: "false"
+  {{- end }}
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply helm/templates/configmap.yaml]: done.

File: src/commands/append/append.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        exitCode = ExitCodes.S3_ERROR;
+      } else if (error instanceof HttpUpstreamUnavailableError) {
+        exitCode = ExitCodes.REMOTE_SERVICE_UNAVAILABLE;
+      } else if (error instanceof HttpUpstreamResponseError) {
+        exitCode = ExitCodes.REMOTE_SERVICE_RESPONSE_ERROR;
+      } else if (error instanceof Osm2pgsqlError) {
+        exitCode = ExitCodes.OSM2PGSQL_ERROR;
+      } else if (error instanceof OsmiumError) {
+        exitCode = ExitCodes.OSMIUM_ERROR;
+      }
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply src/commands/append/append.ts]: done.

File: src/commands/create/create.ts (hunk 1 above)
Comment (melancholiai) [reply src/commands/create/create.ts]: done.

File: src/commands/create/create.ts (hunk 2 above)
Comment (melancholiai) [reply src/commands/create/create.ts]: done.

File: helm/templates/configmap.yaml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  LOG_LEVEL: {{ .Values.env.logLevel | quote }}
+  LOG_PRETTY_PRINT_ENABLED: {{ .Values.env.logPrettyPrintEnabled | quote | default "false" }}
+  {{ if .Values.env.tracing.enabled }}
+  TELEMETRY_TRACING_ENABLED: "true"
+  TELEMETRY_TRACING_URL: {{ $tracingUrl }}
+  {{- else }}
+  TELEMETRY_TRACING_ENABLED: "false"
+  {{- end }}
+  HTTP_CLIENT_TIMEOUT: {{ .Values.env.httpClient.timeout | quote }}
+  COMMAND: {{ .Values.cli.command | quote }}
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply helm/templates/configmap.yaml]: k8s args overrides the defined args from `yargs` and passes only those who are provided in `cronjob.yaml`, in this case only the command itself without its arguments

File: helm/templates/schema-configmap.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{{- $chartName := include "osm2pgsql-wrapper.name" . -}}
+{{- if and (.Values.enabled) (eq .Values.cli.command "append") -}}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-{{ $chartName }}-schema-configmap
+data:
+  {{ .Values.cli.append.config.fileName }}: |-
+{{ .Values.configSchema | indent 4 }}
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply helm/templates/schema-configmap.yaml:9]: done.

File: src/cliBuilderFactory.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+export const cliBuilderFactory: FactoryFunction<Argv> = (dependencyContainer) => {
+  const args = yargs()
+    .env()
+    .usage('Usage: $0 <command> [options]')
+    .demandCommand(1, 'Please provide a command')
+    .option('s3Endpoint', { alias: ['e', 's3-endpoint'], describe: 'The s3 endpoint', nargs: 1, type: 'string', demandOption: true })
+    .option('s3BucketName', {
+      alias: ['b', 's3-bucket-name'],
+      describe: 'The bucket name containing the state and script',
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply src/cliBuilderFactory.ts:24]: done.

File: src/cliBuilderFactory.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+      nargs: 1,
+      type: 'string',
+      demandOption: true,
+    })
+    .option('s3KeyId', {
+      alias: ['k', 's3-key-id'],
+      describe: 'The unique id for the state and script keys',
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply src/cliBuilderFactory.ts:31]: done.

File: src/cliBuilderFactory.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+      nargs: 1,
+      type: 'string',
+      demandOption: true,
+    })
+    .option('s3Acl', { alias: ['a', 's3-acl'], describe: 'The Acl policy for uploaded objects', nargs: 1, type: 'string', default: 'public-read' })
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply src/cliBuilderFactory.ts]: done.

File: src/commands/append/append.ts [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+      await this.manager.getStartSequenceNumber(s3BucketName);
+
+      await this.manager.getEndSequenceNumber(replicationUrl);
+
+      if (this.manager.isUpToDate()) {
+        this.logger.info(`state is up to date, there is nothing to append.`);
+        return;
+      }
+
+      await this.manager.getScripts(s3BucketName);
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply src/commands/append/append.ts:71]: done.

File: src/commands/append/appendManager.ts [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+  SERVICES,
+  DEFAULT_SEQUENCE_NUMBER,
+  SEQUENCE_NUMBER_REGEX,
+  DIFF_TOP_DIR_DIVIDER,
+  DIFF_BOTTOM_DIR_DIVIDER,
+  DIFF_STATE_FILE_MODULO,
+  DIFF_FILE_EXTENTION,
+  STATE_FILE,
+  EXPIRE_LIST,
+  SEQUENCE_NUMBER_PAD_AMOUNT,
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply src/commands/append/appendManager.ts:16]: done.

File: src/commands/append/appendManager.ts [hunk 16]
```diff
@@ -51,84 +52,86 @@ export class AppendManager {
...
+      const scriptFileContent = await streamToString(scriptStream);
+      const localScriptPath = join(DATA_DIR, scriptKey);
+      await createDirectory(getFileDirectory(localScriptPath));
+      await fsPromises.writeFile(localScriptPath, scriptFileContent);
+    })
+
+    await Promise.all(getScriptPromises);
   }
 
   public async appendReplications(replicationUrl: string, bucket: string, acl: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply src/commands/append/appendManager.ts:93]: done.

File: Dockerfile [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+  g++ \
+  libboost-dev \
+  libboost-system-dev \
+  libboost-filesystem-dev \
+  libexpat1-dev \
+  zlib1g-dev \
+  libbz2-dev \
+  libpq-dev \
+  libproj-dev \
+  lua5.3 \
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply Dockerfile]: not needed, done.

File: Dockerfile [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+  liblua5.3-dev \
+  pandoc \
+  git-core \
+  libboost-program-options-dev
+
+RUN git clone git://github.com/openstreetmap/osm2pgsql.git && \
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply Dockerfile]: done.

File: Dockerfile [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  cd osm2pgsql && \
+  mkdir build && \
+  cd build && \
+  cmake .. && \
+  make && \
+  make install
+
+RUN git clone https://github.com/osmcode/osmium-tool && \
+  git clone https://github.com/mapbox/protozero && \
+  git clone https://github.com/osmcode/libosmium && \
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply Dockerfile]: done.

File: Dockerfile [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+RUN npm install
+COPY . .
+RUN npm run build
+
+FROM ubuntu:20.04 as production
+
+ENV DEBIAN_FRONTEND=noninteractive
+ENV workdir /app
+ARG NODE_VERSION=14.x
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply Dockerfile]: done.

**Review by CptSchnitz — APPROVED** [review]

File: src/commands/append/appendManager.ts [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  public async appendReplications(replicationUrl: string, bucket: string, acl: string): Promise<void> {
+    this.logger.info(`${this.appendId} current sequence number ${this.current}`);
+
+    const diffPath = await this.getDiffToFs(replicationUrl);
+
+    const simplifiedDiffPath = await this.simplifyDiff(diffPath);
+
+    const appendPromises = this.entities.map(async (entity) => {
```
[the remark this answers is not in the document]
Comment (melancholiai) [reply src/commands/append/appendManager.ts:106]: no.

**Pull request merged**